lib/pacemaker/pcmk_status.c

```diff
@@ -18,6 +20,246 @@
+                 pcmk_rc_str(rc));
+        pcmk_free_ipc_api(pacemakerd_api);
+        return rc;
+    }
+
+    if (state == pcmk_pacemakerd_state_running || state == pcmk_pacemakerd_state_shutting_down) {
+        rc = pcmk_rc_ok;
+    }
+
+    pcmk_free_ipc_api(pacemakerd_api);
+    return rc;
+}
+
+int
+pcmk__output_cluster_status(pcmk__output_t *out, stonith_t *st, cib_t *cib,
+                            xmlNode *current_cib, pcmk__fence_history_e fence_history,
+                            uint32_t show, uint32_t show_opts, char *only_node,
+                            char *only_rsc, char *neg_location_prefix, bool simple_output)
+{
+    xmlNode *cib_copy = copy_xml(current_cib);
+    stonith_history_t *stonith_history = NULL;
+    int history_rc = 0;
+    pe_working_set_t *data_set = NULL;
+    GList *unames = NULL;
+    GList *resources = NULL;
+
+    int rc = pcmk_rc_ok;
+
+    if (cli_config_update(&cib_copy, NULL, FALSE) == FALSE) {
+        cib__clean_up_connection(&cib);
+        rc = pcmk_rc_schema_validation;
+        out->err(out, "Upgrade failed: %s", pcmk_rc_str(rc));
+        return rc;
+    }
+
+    /* get the stonith-history if there is evidence we need it */
+    if (fence_history != pcmk__fence_history_none) {
+        history_rc = pcmk__get_fencing_history(st, &stonith_history, fence_history);
+    }
+
+    data_set = pe_new_working_set();
+    CRM_ASSERT(data_set != NULL);
+    pe__set_working_set_flags(data_set, pe_flag_no_compat);
+
+    data_set->input = cib_copy;
+    data_set->priv = out;
+    cluster_status(data_set);
+
+    if (pcmk_is_set(show, pcmk_section_bans) || pcmk_is_set(show, pcmk_section_tickets)) {
+        pcmk__unpack_constraints(data_set);
+    }
+
+    unames = pe__build_node_name_list(data_set, only_node);
+    resources = pe__build_rsc_list(data_set, only_rsc);
+
+    /* Always print DC if NULL. */
+    if (data_set->dc_node == NULL) {
+        show |= pcmk_section_dc;
+    }
+
+    if (simple_output) {
+        rc = pcmk__output_simple_status(out, data_set);
+    } else {
+        out->message(out, "cluster-status", data_set, pcmk_rc2exitc(history_rc),
+                     stonith_history, fence_history, show, show_opts,
+                     neg_location_prefix, unames, resources);
+    }
+
+    g_list_free_full(unames, free);
+    g_list_free_full(resources, free);
+
+    stonith_history_free(stonith_history);
+    stonith_history = NULL;
+    pe_reset_working_set(data_set);
+    return pcmk_rc_ok;
+}
+
+int
+pcmk__status(pcmk__output_t *out, cib_t *cib, pcmk__fence_history_e fence_history,
+             uint32_t show, uint32_t show_opts, char *only_node, char *only_rsc,
+             char *neg_location_prefix, bool simple_output)
+{
+    xmlNode *current_cib = NULL;
+    int rc = pcmk_rc_ok;
+    stonith_t *st = NULL;
+
+    rc = pacemakerd_status(out, cib, cib->variant == cib_native);
+    if (rc != pcmk_rc_ok) {
+        return rc;
```

Comment:
I think it's questionable whether we should abort on connection failures, rather than still show whatever we can. But since that's how it works currently, we can leave it alone